### ISSUE TITLE
This call to matplotlib.use() has no effect because the backend has already

### ISSUE THREAD

**Issue by iweey** [JosPolfliet/pandas-profiling#68]
/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/pandas_profiling/base.py:20: UserWarning: 
This call to matplotlib.use() has no effect because the backend has already
been chosen; matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.

The backend was *originally* set to 'module://ipykernel.pylab.backend_inline' by the following code:
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/ipykernel/kernelapp.py", line 477, in start
    ioloop.IOLoop.instance().start()
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/zmq/eventloop/ioloop.py", line 177, in start
    super(ZMQIOLoop, self).start()
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/tornado/ioloop.py", line 888, in start
    handler_func(fd_obj, events)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/tornado/stack_context.py", line 277, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 440, in _handle_events
    self._handle_recv()
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 472, in _handle_recv
    self._run_callback(callback, msg)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 414, in _run_callback
    callback(*args, **kwargs)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/tornado/stack_context.py", line 277, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/ipykernel/kernelbase.py", line 283, in dispatcher
    return self.dispatch_shell(stream, msg)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/ipykernel/kernelbase.py", line 235, in dispatch_shell
    handler(stream, idents, msg)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/ipykernel/kernelbase.py", line 399, in execute_request
    user_expressions, allow_stdin)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/ipykernel/ipkernel.py", line 196, in do_execute
    res = shell.run_cell(code, store_history=store_history, silent=silent)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/ipykernel/zmqshell.py", line 533, in run_cell
    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2718, in run_cell
    interactivity=interactivity, compiler=compiler, result=result)
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2822, in run_ast_nodes
    if self.run_code(code, result):
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2882, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-47d9d0ad501d>", line 8, in <module>
    import matplotlib.pyplot as plt
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/matplotlib/pyplot.py", line 69, in <module>
    from matplotlib.backends import pylab_setup
  File "/home/flash1/work/software/python/anaconda2/lib/python2.7/site-packages/matplotlib/backends/__init__.py", line 14, in <module>
    line for line in traceback.format_stack()

  matplotlib.use('Agg')

**Comment by romainx** [comment]
Hello,

Could you please explain your use case and take the time to format the issue?

Many thanks

**Comment by ianozsvald** [comment]
+1 I have this issue. I've narrowed it down to:
```
import seaborn
import pandas_profiling
# only these two lines will generate:
/home/ian/anaconda3/envs/data_science_delivered/lib/python3.6/site-packages/pandas_profiling/base.py:20: UserWarning: 
This call to matplotlib.use() has no effect because the backend has already
been chosen; matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.

The backend was *originally* set to 'module://ipykernel.pylab.backend_inline' by the following code:
  File "/home/ian/anaconda3/envs/data_science_delivered/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
...
    from matplotlib.backends import pylab_setup
  File "/home/ian/anaconda3/envs/data_science_delivered/lib/python3.6/site-packages/matplotlib/backends/__init__.py", line 14, in <module>
    line for line in traceback.format_stack()

  matplotlib.use('Agg')
```

My versions (using `watermark`):
```
2017-11-15 

CPython 3.6.2
IPython 6.1.0

numpy 1.13.3
matplotlib 2.1.0
seaborn 0.8.0

compiler   : GCC 4.8.2 20140120 (Red Hat 4.8.2-15)
system     : Linux
release    : 4.9.45-040945-generic
machine    : x86_64
processor  : x86_64
CPU cores  : 8
interpreter: 64bit
```

If I invert those lines (so seaborn is second) then the warning goes away, if I do either combo in my _real_ import list then I still get the huge warning:
```
import datetime
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import pandas_profiling
import seaborn as sns
import tqdm

import eli5 # https://github.com/TeamHG-Memex/eli5 via conda
import lime # https://github.com/marcotcr/lime via PyPI (note this was built using MASTER at time-of-creation)
import yellowbrick # https://github.com/DistrictDataLabs/yellowbrick/ via conda
from eli5 import show_weights
from eli5 import show_prediction

import lime.lime_tabular
from matplotlib import pyplot as plt
%matplotlib inline```

**Comment by ianozsvald** [comment]
I'll note that if comment-out `matplotlib.use('Agg')` in `base.py` - this problem goes away. Maybe that's the easy fix?

I _think_ in my Notebook that the line is redundant as I'm using the Jupyter backend, my first few lines are:
```
import matplotlib
print(matplotlib.get_backend())
-> module://ipykernel.pylab.backend_inline
```
and when I subsequently import `pandas_profiling` the backend doesn't change (with the Agg line enabled in `base.py`).

@iweey are you using a Jupyter Notebook too? Your error looks to be the same as mine?

**Comment by iweey** [comment]
@ianozsvald  yes,Jupyter Notebook meet this issue

**Comment by ianozsvald** [comment]
I'll note that I've ignored this warning using:
```
import warnings
warnings.filterwarnings('ignore') 
```
over here: https://github.com/ianozsvald/simple_generative_insurance_model/blob/master/simple_generative_insurance_model.ipynb

This is a blunt-stick approach, commenting out `matplotlib.use('Agg')` in `base.py` works fine too and I do wonder if that needs to be set by `pandas_profiling` as it feels user-machine-specific?

**Comment by romainx** [comment]
Hello, 

Thanks to all for your comments ! Here is a [link](http://matplotlib.org/faq/usage_faq.html#what-is-a-backend) to the reference documentation on backend.

> If you use the use() function, this must be done before importing `matplotlib.pyplot`. Calling `use()` after `pyplot` has been imported will have no effect. Using `use()` will require changes in your code if users want to use a different backend. Therefore, you should avoid explicitly calling `use()` unless absolutely necessary.

So you may be right, the explicit call to `use` may be useless. However my guess is that it has been made to ensure that figures in the report are generated in PNG format. 
I will dig in to find if it is necessary in all contexts, I have also to check the behavior when the tool is run outside Jupyter (call to `profile_csv.py`). And if so, maybe there is a better way to handle it when it's already set.

Best.

**Comment by conradoqg** [comment]
Hey guys!

I commented the `matplotlib.use('Agg')` line and tested the `profile_csv.py` and jupyter notebook. Everything seems to work without it. I think we should remove it, I can do an PR if necessary.

Best.

**Comment by romainx** [comment]
Hello,

Many thanks for the test, I have commented the line and merged the change #80.
Really appreciate your contribution.

**Comment by ianozsvald** [comment]
Hi @romainx, I don't think that this fix is right - having upgraded to 1.4.1 I still have the same issue as reported above. Specifically in a fresh Notebook, I see:
```
import matplotlib
matplotlib.get_backend()
-> 'module://ipykernel.pylab.backend_inline'
```

The proposed fix ( https://github.com/pandas-profiling/pandas-profiling/blob/master/pandas_profiling/plot.py#L12 ) checks if the backed == `agg` and if not, it forces to `agg`. 

If I comment out the `if` statement (so the backend isn't changed) then the Warning goes away.

I don't know what the case is if you are just running the tests, but forcing Agg even though the user already has another renderer isn't the right way to go. Is the renderer something like a blank string when the tests run?

**Comment by aagostini-tada** [comment]
I've seen the same issue

**Comment by conradoqg** [comment]
I'm gonna check this.

**Comment by sleitner** [comment]
Was any progress made here? The unit tests hang for me, and I traced it back to this backend issue, when my default interactive matplotlib backend is used (`TkAgg` -- from my personal matplotlibrc) *and* the multiprocessing `pool_size>1`.

I agree that it's heavy handed to force the user to `matplotlib.use('Agg')` even when `pool_size=1`, but I would have appreciated a warning not to use an interactive backend, especially given that the default pool_size in `describe.py` is naturally going to be `>1` for most everyone.

**Comment by anilsomisetty** [comment]
I am getting an error like this Can someone help me with this 

utils/visualization_utils.py:27: UserWarning: 
This call to matplotlib.use() has no effect because the backend has already
been chosen; matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.

The backend was *originally* set to 'module://ipykernel.pylab.backend_inline' by the following code:
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/anil/.virtualenvs/ml1/lib/python2.7/site-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/ipykernel/kernelapp.py", line 499, in start
    self.io_loop.start()
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/tornado/ioloop.py", line 1017, in start
    self._run_callback(self._callbacks.popleft())
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/tornado/ioloop.py", line 758, in _run_callback
    ret = callback()
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/tornado/stack_context.py", line 300, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 536, in <lambda>
    self.io_loop.add_callback(lambda : self._handle_events(self.socket, 0))
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 450, in _handle_events
    self._handle_recv()
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 480, in _handle_recv
    self._run_callback(callback, msg)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 432, in _run_callback
    callback(*args, **kwargs)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/tornado/stack_context.py", line 300, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/ipykernel/kernelbase.py", line 283, in dispatcher
    return self.dispatch_shell(stream, msg)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/ipykernel/kernelbase.py", line 233, in dispatch_shell
    handler(stream, idents, msg)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/ipykernel/kernelbase.py", line 399, in execute_request
    user_expressions, allow_stdin)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/ipykernel/ipkernel.py", line 208, in do_execute
    res = shell.run_cell(code, store_history=store_history, silent=silent)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/ipykernel/zmqshell.py", line 537, in run_cell
    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2714, in run_cell
    interactivity=interactivity, compiler=compiler, result=result)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2824, in run_ast_nodes
    if self.run_code(code, result):
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2878, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-48fe9407af3a>", line 2, in <module>
    get_ipython().magic(u'matplotlib inline')
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2160, in magic
    return self.run_line_magic(magic_name, magic_arg_s)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2081, in run_line_magic
    result = fn(*args,**kwargs)
  File "<decorator-gen-105>", line 2, in matplotlib
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/magic.py", line 188, in <lambda>
    call = lambda f, *a, **k: f(*a, **k)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/magics/pylab.py", line 100, in matplotlib
    gui, backend = self.shell.enable_matplotlib(args.gui)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2966, in enable_matplotlib
    pt.activate_matplotlib(backend)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/IPython/core/pylabtools.py", line 315, in activate_matplotlib
    matplotlib.pyplot.switch_backend(backend)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/matplotlib/pyplot.py", line 231, in switch_backend
    matplotlib.use(newbackend, warn=False, force=True)
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/matplotlib/__init__.py", line 1422, in use
    reload(sys.modules['matplotlib.backends'])
  File "/home/anil/.virtualenvs/ml1/local/lib/python2.7/site-packages/matplotlib/backends/__init__.py", line 16, in <module>
    line for line in traceback.format_stack()


  import matplotlib; matplotlib.use('Agg')  # pylint: disable=multiple-statements
